lib/mail_catcher.rb

```diff
@@ -191,80 +184,84 @@ def run! options=nil
     # Stash them away for later
     @@options = options
 
-    # If we're running in the foreground sync the output.
-    unless options[:daemon]
-      $stdout.sync = $stderr.sync = true
-    end
+    # sync the output.
+    $stdout.sync = $stderr.sync = true
 
     puts "Starting MailCatcher"
 
-    Thin::Logging.debug = development?
-    Thin::Logging.silent = !development?
+    Async.logger.level = Logger::DEBUG if options[:verbose]
 
-    # One EventMachine loop...
-    EventMachine.run do
-      # Set up an SMTP server to run within EventMachine
-      rescue_port options[:smtp_port] do
-        EventMachine.start_server options[:smtp_ip], options[:smtp_port], Smtp
-        puts "==> #{smtp_url}"
-      end
+    Async::Reactor.run do |task|
+      smtp_address = Async::IO::Address.tcp(options[:smtp_ip], options[:smtp_port])
+      smtp_endpoint = Async::IO::AddressEndpoint.new(smtp_address)
+      smtp_socket = rescue_port(options[:smtp_port]) { smtp_endpoint.bind }
+      puts "==> #{smtp_url}"
 
-      # Let Thin set itself up inside our EventMachine loop
-      # (Skinny/WebSockets just works on the inside)
-      rescue_port options[:http_port] do
-        Thin::Server.start(options[:http_ip], options[:http_port], Web)
-        puts "==> #{http_url}"
+      smtp_endpoint = MailCatcher::SMTP::URLEndpoint.new(URI.parse(smtp_url), smtp_endpoint)
+      smtp_server = MailCatcher::SMTP::Server.new(smtp_endpoint) do |envelope|
+        MailCatcher::Mail.add_message(sender: envelope.sender, recipients: envelope.recipients,
+                                      source: envelope.content)
       end
 
-      # Open the web browser before detatching console
-      if options[:browse]
-        EventMachine.next_tick do
-          browse http_url
+      smtp_task = task.async do |task|
+        task.annotate "binding to #{smtp_socket.local_address.inspect}"
+
+        begin
+          smtp_socket.listen(Socket::SOMAXCONN)
+          smtp_socket.accept_each(task: task, &smtp_server.method(:accept))
+        ensure
+          smtp_socket.close
         end
       end
 
-      # Daemonize, if we should, but only after the servers have started.
-      if options[:daemon]
-        EventMachine.next_tick do
-          if quittable?
-            puts "*** MailCatcher runs as a daemon by default. Go to the web interface to quit."
-          else
-            puts "*** MailCatcher is now running as a daemon that cannot be quit."
-          end
-          Process.daemon
```

Comment:
Why was daemonisation removed?

Reply:
Adding it in the next commit. Keeping it in the async block was blocking the process instead of daemonisation. I got the solution for this 😅.